lib/CSV/HistoryPlayer.pm

```diff
@@ -0,0 +1,134 @@
+package CSV::HistoryPlayer;
+
+use List::MoreUtils qw(uniq);
+use Moo;
+use Path::Tiny;
+use Text::CSV;
+
+use strictures 2;
+use namespace::clean;
+
+our $VERSION = '0.01';
+
+has 'root_dir' => (
+    is       => 'ro',
+    required => 1
+);
+
+has 'dir_filter' => (
+    is      => 'ro',
+    default => sub {
+        return sub { 1 }
+    });
+
+has 'files_maper' => (
+    is      => 'ro',
+    default => sub {
+        return sub {
+            my $files = shift;
+            return [sort { $a cmp $b } @$files];
+            }
+    });
+
+has 'files' => (is => 'lazy');
+
+has _current_data => (is => 'rw');
+
+has '_reader' => (is => 'lazy');
+
+sub _build_files {
+    my $self = shift;
+
+    my @files;
+    my @dirs_queue = (path($self->root_dir));
+    my $dir_filter = $self->dir_filter;
+    while (@dirs_queue) {
+        my $dir = shift(@dirs_queue);
+        if ($dir_filter->($dir)) {
+            for my $c ($dir->children) {
+                push @dirs_queue, $c if (-d $c);
+                push @files, $c
+                    if ($c =~ /\.csv$/i && -s $c && -r $c && -f $c);
+            }
+        }
+    }
+    my $sorted_files = $self->files_maper->(\@files);
+    return $sorted_files;
+}
+
+sub _build__reader {
+    my $self        = shift;
+    my $files       = $self->files;
+    my $clusters    = [uniq map { $_->basename } @$files];
+    my $cluster_idx = -1;
+    my @cluster_fds;
+    my @cluster_csvs;
+    my @cluser_files;
+
+    my $open_cluster = sub {
+        my $cluster_id = $clusters->[$cluster_idx];
+        @cluser_files = grep { $_->basename eq $cluster_id } @$files;
+        @cluster_fds  = ();
+        @cluster_csvs = ();
+        for my $cf (@cluser_files) {
+            my $csv = Text::CSV->new({binary => 1})
+                or die "Cannot use CSV: " . Text::CSV->error_diag();
+            my $fh = $cf->filehandle("<");
+            push @cluster_fds,  $fh;
+            push @cluster_csvs, $csv;
+        }
+    };
+
+    my @lines;
+    my $read_line_from_cluster = sub {
+        REDO:
+
+        # make sure that we read last line from all cluster files
+        for my $idx (0 .. @cluster_fds - 1) {
+            if (!defined $lines[$idx] && !$cluster_csvs[$idx]->eof) {
+                $lines[$idx] =
+                    $cluster_csvs[$idx]->getline($cluster_fds[$idx]);
+            }
+        }
+
+        # we assume that timestamp is the 1st column
+        my @ordered_idx =
+            sort { $lines[$a]->[0] <=> $lines[$b]->[0] }
+            grep { defined $lines[$_] } (0 .. @lines - 1);
+        if (@ordered_idx) {
+            my $idx  = shift @ordered_idx;
+            my $line = $lines[$idx];
+            my $file = $cluser_files[$idx];
+            $self->_current_data([$file, $line]);
+            $lines[$idx] = undef;
+        } else {
+            if ($cluster_idx < @$clusters - 1) {
+                $open_cluster->(++$cluster_idx);
+                goto REDO;
```

Comment:
Just a style issue, could you please convert this to something like:

```perl
REDO: {
    if () {
    } else {
        ...;
        redo REDO;
    }
}
```

or

```perl
my $read_line_from_cluster;
$read_line_from_cluster = sub {
    ....
    goto &$read_line_from_cluster;
};
```

Yet, I'd prefer the former. I think the latter can create a memory leak.